gunrock/app/lp/lp_problem.cuh

```diff
@@ -0,0 +1,511 @@
+      data.SetName("data_size");
+
+      segments.SetName("segments");
+      segments.SetName("segments_size");
+      cub_temp_storage.SetName("cub_temp_storage");
+      segments_temp.SetName("segments_temp");
+    }
+
+    /*
+     * @brief Default destructor
+     */
+    virtual ~DataSlice() { Release(); }
+
+    cudaError_t Release(util::Location target = util::LOCATION_ALL) {
+      cudaError_t retval = cudaSuccess;
+      if (target & util::DEVICE) GUARD_CU(util::SetDevice(this->gpu_idx));
+
+      // GUARD_CU(original_vertex      .Release(target));
+      GUARD_CU(labels.Release(target));
+      GUARD_CU(old_labels.Release(target));
+      GUARD_CU(preds.Release(target));
+      GUARD_CU(temp_preds.Release(target));
+      GUARD_CU(vertex_markers[0].Release(target));
+      GUARD_CU(vertex_markers[1].Release(target));
+      GUARD_CU(unvisited_vertices[0].Release(target));
+      GUARD_CU(unvisited_vertices[1].Release(target));
+      GUARD_CU(split_lengths.Release(target));
+      GUARD_CU(local_vertices.Release(target));
+      GUARD_CU(visited_masks.Release(target));
+      GUARD_CU(old_mask.Release(target));
+      GUARD_CU(in_masks.Release(target));
+      GUARD_CU(in_masks.Release(target));
+      GUARD_CU(data.Release(target));
+      GUARD_CU(data_size.Release(target));
+      GUARD_CU(segments_size.Release(target));
+      GUARD_CU(segments.Release(target));
+      GUARD_CU(segments_temp.Release(target));
+      GUARD_CU(cub_temp_storage.Release(target));
+      GUARD_CU(BaseDataSlice ::Release(target));
+      return retval;
+    }
+
+    /**
+     * @brief initializing sssp-specific data on each gpu
+     * @param     sub_graph   Sub graph on the GPU.
+     * @param[in] num_gpus    Number of GPUs
+     * @param[in] gpu_idx     GPU device index
+     * @param[in] target      Targeting device location
+     * @param[in] flag        Problem flag containling options
+     * \return    cudaError_t Error message(s), if any
+     */
+    cudaError_t Init(GraphT &sub_graph, int num_gpus = 1, int gpu_idx = 0,
+                     util::Location target = util::DEVICE,
+                     ProblemFlag flag = Problem_None) {
+      cudaError_t retval = cudaSuccess;
+
+      GUARD_CU(BaseDataSlice::Init(sub_graph, num_gpus, gpu_idx, target, flag));
+
+      GUARD_CU(labels.Allocate(sub_graph.nodes, target));
+      GUARD_CU(old_labels.Allocate(sub_graph.nodes, target));
+      GUARD_CU(segments.Allocate(sub_graph.nodes, target));
+      GUARD_CU(segments_temp.Allocate(sub_graph.nodes, target));
+      GUARD_CU(segments_size.Allocate(1, target));
+      GUARD_CU(data_size.Allocate(1,  util::DEVICE | util::HOST));
+      GUARD_CU(cub_temp_storage.Allocate(1, target));
+      
+      // TODO
+      // all the remaining space should be taken by data
+      // can we preallocate something like that?
+      // we cannot currently dynamically allocate space
+      // so the maximum is the best but then cuda malloc calls can be slow
+      
```

Comment:
What do you mean by this? If a user has a card with 12GB of storage, and we are currently using 1GB normally, we should allocate an additional 11GB? That is something that we can do, but why would we do this?